GLDS: Create AutoSaveField component (#61316)
grafana/grafana: packages/grafana-ui/src/components/AutoSaveField/AutoSaveField.mdx

import { Props, Preview } from '@storybook/addon-docs/blocks';
import { AutoSaveField } from './AutoSaveField';
import { Basic } from './AutoSaveField.story.tsx';

# AutoSaveField

> Used for text input that saves automatically its content. For an array of data or tree-structured data, consider using `Select` or `Cascader` respectively.

Reviewer: ```suggestion
Used for form inputs that should save its content automatically. 
```

Also do we want to add what is at the top of the autosave doc as a sort of "warning" to use this sparingly?

Author: If we want a "warning" I think this is the best place to show it.

## <a name="usage"/> Usage

<br />

### **Do's**

- Use in **low-impact** areas within Grafana (examples; Team name, display name)
- Include helper text to tell users of the new behaviour for recently modified elements

<br />

### **Dont's**

- Do not use if high-impact situations (e.g. deleting a data source)
  <br />
  <br />

## <a name="variants"/>Variants

### 1. Text:

- The `Input` component allows saving the value of a string.
- This is the main component thought to work as an `AutoSaveField` child.

```jsx
<AutoSaveField
  onFinishChange={customRequest}
  //Complete field args if needed
>
  {(onChange) => (
    <Input
      value={inputTextValue}
      onChange={(e) => {
        onChange(e.currentTarget.value);
        //Complete code if needed
      }}
    />
  )}
</AutoSaveField>
```

### 2. TextArea:

The `TextArea` component allows saving a multi-line text.

```jsx
<AutoSaveField
  onFinishChange={customRequest}
  //Complete field args if needed
>
  {(onChange) => (
    <TextArea
      value={textAreaValue}
      onChange={(e) => {
        onChange(e.currentTarget.value);
        //Complete code if needed
      }}
    />
  )}
</AutoSaveField>
```

### 3. Checkbox:

> The `Checkbox` component allows saving an specific value clicking or not the checkbox input.

Reviewer: ```suggestion
The `Checkbox` component allows saving a specific value by ticking or not the checkbox input.
```

2. **Input element:** this component must have an input as a child. There is a list of tested components in the ‘Variants’ section.
3. **InlineToast:** when the input value is successfully saved, a toast notification with the text ‘Saved!’ will appear next to the proper field.
4. **Error message:** this element will appear when the request fails. This message can be customized by the user or be the response of the request.
   <br />

### Optional elements:

- **Description:** used to inform the user the value of this input will be save automatically. _This is tentative requirement until users become familiar with this pattern_
  <br />
  <br />

## <a name="formating"/>Formating

<br />

### Anatomy

The `AutoSaveField`component is comprised of:

- A slot for form elements (e.g text input, checkbox, dropdown etc)
- Inline toast notification: used to indicate temporal status near fields/components, such as a Saved! indicator next to a field, or a little Copied! indicator above a button
- Inline validation
- Helper text: used to inform the user the value of this input will be save automatically
- Custom error message

{' '}
<Preview>
  <Basic />

> </Preview>

Author: @staton-hyse11 can you create the `AutoSaveField` image for the Anatomy section?

import { Props, Preview } from '@storybook/addon-docs/blocks';
import { AutoSaveField } from './AutoSaveField';
import { Basic } from './AutoSaveField.story.tsx';

# AutoSaveField

Used for form inputs that should save its content automatically.

---

In this documentation you can find:

1. [Usage](#usage)
1. [Variants](#variants)

> 1. [Accesibility](#accessibility)

Reviewer: ```suggestion
1. [Accessibility](#accessibility)
```

1. [Content](#content)

> 1. [Formating](#formating)

Reviewer: ```suggestion
1. [Formatting](#formatting)
```

        onChange(value);
        //Complete code if needed
      }}
    />
  )}
</AutoSaveField>
```

<br />
<br />

## <a name="accessibility"/>Accessibility

`AutoSaveField` has inherited the accessibility properties of both the `Field` and the `Input` component it is comprised of.
Apart from that, the `InlineToast`, used to show the success of the request, has an [aria-live](https://www.w3.org/TR/wai-aria/#aria-live) property set to polite, so assistive technologies will notify users of updates but generally do not interrupt the current task.

<br />
<br />

## <a name="content"/>Content

<br />

### Main elements

<br />

1. **Label:** the AutoSaveField must be accompanied with labels that should be clear and descriptive.
2. **Input element:** this component must have an input as a child. There is a list of tested components in the ‘Variants’ section.
3. **InlineToast:** when the input value is successfully saved, a toast notification with the text ‘Saved!’ will appear next to the proper field.
4. **Error message:** this element will appear when the request fails. This message can be customized by the user or be the response of the request.
   <br />

### Optional elements:

> - **Description:** used to inform the user the value of this input will be save automatically. _This is tentative requirement until users become familiar with this pattern_

Reviewer: ```suggestion
- **Description:** used to inform the user the value of this input will be saved automatically. _This is a tentative requirement until users become familiar with this pattern_
```

  <br />
  <br />

> ## <a name="formating"/>Formating

Reviewer: ```suggestion
## <a name="formatting"/>Formatting
```
